src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+- Catboost: Well calibrated with slightly under-predictions for mid-high ranges
+
+I had a doubt about why predicted probability stops at 0.6 for XGBoost, so wanted to do a quick checks on that. Let's plot only the XGBoost modifications, to see if the model gains a bit of probability when it becomes more complex:
+
+
+```python
+# Best XGBoost model based on the RandomizedSearchCV results
+high_estimator_xgb_model = xgb.XGBClassifier(
+    n_estimators=250,
+    learning_rate=0.05,
+    max_depth=3,
+    reg_lambda=0.5,  # 'lambda' is renamed to 'reg_lambda' in XGBClassifier
+    objective='binary:logistic',
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+high_estimator_xgb_model.fit(X_train, y_train)
+y_pred_high_estimator_xgb_model = high_estimator_xgb_model.predict_proba(X_val)[:, 1]
+
+# Best XGBoost model based on the RandomizedSearchCV results
+deeper__xgb_model = xgb.XGBClassifier(
+    n_estimators=50,
+    learning_rate=0.05,
+    max_depth=6,
+    reg_lambda=0.5,  # 'lambda' is renamed to 'reg_lambda' in XGBClassifier
+    objective='binary:logistic',
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+deeper__xgb_model.fit(X_train, y_train)
+y_pred_deeper__xgb_model = deeper__xgb_model.predict_proba(X_val)[:, 1]
+
+# Best XGBoost model based on the RandomizedSearchCV results
+high_deeper__xgb_model = xgb.XGBClassifier(
+    n_estimators=250,
+    learning_rate=0.05,
+    max_depth=6,
+    reg_lambda=0.5,  # 'lambda' is renamed to 'reg_lambda' in XGBClassifier
+    objective='binary:logistic',
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+high_deeper__xgb_model.fit(X_train, y_train)
+y_pred_high_deeper__xgb_model = high_deeper__xgb_model.predict_proba(X_val)[:, 1]
+
+# Evaluate
+generate_evaluation_curves_multi(
+    model_names=["LogisticRegression"] + ["GBT(n_trees=100, depth=3)"] + ["XGBoost (n_trees=50, depth=3, lr=0.05)"] + ["High-estimator"] + ["Deeper"]+ ["High&Deeper"],
+    y_preds=[y_pred_lr, y_pred_gbt,y_pred_best_xgb, y_pred_high_estimator_xgb_model, y_pred_deeper__xgb_model, y_pred_high_deeper__xgb_model ],
+    y_test=y_val
+)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_52_0.png)
+    
+
+
+
+```python
+plot_calibration_curve(
+    y_val,
+    [y_pred_best_xgb, y_pred_high_estimator_xgb_model, y_pred_deeper__xgb_model, y_pred_high_deeper__xgb_model],
+    ["XGBoost (n_trees=50, depth=3, lr=0.05)", "XGB but n_trees=250", "XGB but depth=6", "XGB (n_trees=250, depth=6)"]
+)
+
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_53_0.png)
+    
+
+
+Indeed they do, my gut-feeling it is that we are providing more complexity to the model, hence letting it decide with more robustness for positive outcomes (although they are not our best model as we saw previously).
```

Comment:
Good observation and nice you went deeper into this. I think your gut feeling is not very far away from reality. To understand what's going on, bear in mind that to emit a probability in a certain leaf, the model is looking into all the samples falling into that leaf and looking at the proportion of them which are 1s. If we have a single tree with no branches and one top leaf, the emited probability should be the proportion of 1s in the whole dataset (which is small). As we add more branches (or more trees) the model starts "overfitting" and grouping a larger proportion of ones into the same leaf, increasing the emited probabilities for them